### PR TITLE
Normative: New pattern matching design

### PR DESCRIPTION
**Pull request by Jack-Works**
This PR contains a new spec that follows #293

https://raw.githack.com/tc39/proposal-pattern-matching/new-spec/index.html

### CODE REVIEW

**Comment by Jack-Works** [comment]
This PR is now ready. I have no more things to change before I receive any input. The Scope Analysis is left as a TODO and it won't be in this PR.

File: spec.emu [hunk 1]
```diff
@@ -1720,42 +1635,2123 @@ contributors:
...
+        Literal
+        NoSubstitutionTemplate
+        PatternMatchingMemberExpression[?Yield, ?Await]
+        PatternMatchingUnaryAlgebraicExpression[?Yield, ?Await]
+    </emu-grammar>
+    <emu-note code>
+      <pre><code class="javascript">
+        if (expr is &gt; -10 and &lt; 10) {}
+        if (expr is { value: instanceof Error and { let message } }) {}
+        if (expr is { type: "object", value: has #field }) {}
```
Comment (gibson042) [inline spec.emu:1987]: `has` is described below as a _possible_ addition.
```suggestion
```

File: spec.emu [hunk 2]
```diff
@@ -1720,42 +1635,2123 @@ contributors:
...
+        `>=` PatternMatchingOrderExpression[?Yield, ?Await]
+        `instanceof` PatternMatchingMemberExpression[?Yield, ?Await]
+        `in` PatternMatchingMemberExpression[?Yield, ?Await]
+        `==` PatternMatchingEqualExpression[?Yield, ?Await]
+        `!=` PatternMatchingEqualExpression[?Yield, ?Await]
+        `===` PatternMatchingEqualExpression[?Yield, ?Await]
+        `!==` PatternMatchingEqualExpression[?Yield, ?Await]
+
+      PatternMatchingOrderExpression[Yield, Await] :
+        NumericLiteral
```
Comment (gibson042) [inline spec.emu:1973]: Support for `> 0` but not `> "m"`?

File: spec.emu [hunk 3]
```diff
@@ -1720,42 +1635,2123 @@ contributors:
...
+    <!-- endregion -->
+
+    <emu-clause id="sec-match-patterns-static-semantics-early-errors">
+      <h1>Static Semantics: Early Errors</h1>
+      <!-- #region PrimitivePattern -->
+      <emu-grammar>
+        PrimitivePattern : NoSubstitutionTemplate
+      </emu-grammar>
+      <ul>
+        <li>It is a Syntax Error |NoSubstitutionTemplate| Contains |NotEscapeSequence|.</li>
```
Comment (gibson042) [inline spec.emu:2134]:
```suggestion
        <li>It is a Syntax Error if |NoSubstitutionTemplate| Contains |NotEscapeSequence|.</li>
```

File: spec.emu [hunk 4]
```diff
@@ -1720,42 +1635,2123 @@ contributors:
...
+      </ul>
+      <!-- #endregion -->
+      <!-- #region ObjectPattern and ArrayPattern -->
+      <emu-grammar>
+        MatchProperty :
+          PropertyName `?`?
+          PropertyName `?`? `:` MatchPattern
+      </emu-grammar>
+      <ul>
+        <li>It is a Syntax Error if |PropertyName| is a |LiteralPropertyName| and PropName of |LiteralPropertyName| is *"__proto__"*.</li>
```
Comment (gibson042) [inline spec.emu:2152]:
```suggestion
        <li>It is a Syntax Error if |PropertyName| is a |LiteralPropertyName| and PropName of |PropertyName| is *"__proto__"*.</li>
```

File: spec.emu [hunk 5]
```diff
@@ -1720,42 +1635,2123 @@ contributors:
...
+      <emu-alg>
+        1. Let _key_ be ? ToPropertyKey(_subject_).
+        1. Let _expr_ be the |Expression| that is covered by |PatternMatchingMemberExpression|.
+        1. Let _ref_ be ? Evaluation of _expr_.
+        1. Let _value_ be ? GetValue(_ref_).
+        1. If _value_ is not an Object, return *false*.
+        1. Return ? HasProperty(_value_, _key_).
+      </emu-alg>
+      <emu-note type="editor">
+        The current proposal caches all operations on the _subject_ except this. Should we cache the result of ToPropertyKey?
```
Comment (gibson042) [inline spec.emu:2954]: Yes! Caching behavior is one of the most confusing introductions here; it **must** be uniform.

File: spec.emu [hunk 6]
```diff
@@ -1720,42 +1635,2123 @@ contributors:
...
+        const regex = /^starts/;
+        if (value is regex) {}
+      </code></pre>
+    </emu-note>
+    <!-- endregion -->
+    <!-- #region MemberExpressionPattern -->
+    <emu-grammar type="definition">
+      MemberExpressionPattern[Yield, Await] :
+        PatternMatchingMemberExpression[?Yield, ?Await]
+        PatternMatchingMemberExpression[?Yield, ?Await] `(` MatchList[?Yield, ?Await] `)`
```
[the remark this answers is not in the document]
Comment (Jack-Works) [reply spec.emu:1772]: this is the extractor syntax we have, it is a feature to visually match a call expression

File: spec.emu (hunk 2 above)
Comment (Jack-Works) [reply spec.emu:1987]: yes, but we can add it. I know there is valid use cases for it but I think maybe it is not a thing we should encourage.

File: spec.emu (hunk 5 above)
Comment (Jack-Works) [reply spec.emu]: added!

File: spec.emu (hunk 6 above)
[the remark this answers is not in the document]
Comment (gibson042) [reply spec.emu:1772]: Visual similarity (or in this case, exact overlap) is a benefit when coupled with behavioral similarity but a _drawback_ when not. For `x is a.b.c` to perform all the expected property lookups and then serve as a matcher but `x is a.b.c()` and `x is a.b.c(1)` to also perform the lookups and then **not** invoke the `c` value to _produce_ such a matcher is counterintuitive and confusing, even if that particular confusion would be absent from `x is a.b.c(let arg)` and `x is a.b.c(instanceof Array)`.

File: spec.emu (hunk 2 above)
Comment (ljharb) [reply spec.emu:1987]: It should match the > operator, including valueOf coercion, i think.

Comment (Jack-Works) [reply spec.emu:1987]: ah, do we really want that? to coercion here?

Comment (ljharb) [reply spec.emu:1987]: It’s irrelevant if we want it; that’s what > does, and must do everywhere.

**Comment by Jack-Works** [comment]
Updated according to decisions made in https://github.com/tc39/proposal-pattern-matching/issues/311

**Pull request merged**